### PR TITLE
Add specific stream protocol counters to track protocol errors

### PR DESCRIPTION
**Pull request by dcorbacho**
## Types of Changes
- [ ] Bug fix (non-breaking change which fixes issue #NNNN)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)

## Checklist
- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [x] All tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

### CODE REVIEW

**Comment by gerhard** [comment]
This can now be rebased onto `master`, #3136 is in. Thank you @michaelklishin 👍🏻

File: deps/rabbitmq_stream/include/rabbit_stream_metrics.hrl [hunk 1]
```diff
@@ -1,2 +1,94 @@
...
+-define(FRAME_TOO_LARGE, ?NUM_PROTOCOL_COUNTERS + 13).
+-define(INTERNAL_ERROR, ?NUM_PROTOCOL_COUNTERS + 14).
+-define(ACCESS_REFUSED, ?NUM_PROTOCOL_COUNTERS + 15).
+-define(PRECONDITION_FAILED, ?NUM_PROTOCOL_COUNTERS + 16).
+-define(PUBLISHER_DOES_NOT_EXIST, ?NUM_PROTOCOL_COUNTERS + 17).
+
+-define(PROTOCOL_COUNTERS,
+        [
+         {
+          stream_error_stream_does_not_exist, ?STREAM_DOES_NOT_EXIST, counter,
```
Comment (gerhard) [inline deps/rabbitmq_stream/include/rabbit_stream_metrics.hrl:27]: If these are counters, then Prometheus format requires them to end with `_total` : 

> Note that an accumulating count has total as a suffix, in addition to the unit if applicable.

https://prometheus.io/docs/practices/naming/

**Comment by gerhard** [comment]
Apart from the above `_total` change, these items are missing before we can merge:

- [ ] modify `deps/rabbit/queue_type_SUITE > smoke` to check for these new metrics
- [ ] modify `deps/rabbitmq_prometheus/rabbit_prometheus_http_SUITE.erl > global_metrics_present_test` to check for these new metrics
- [ ] add metric to `deps/rabbitmq_prometheus/metrics.md` under GLOBAL COUNTERS

@Gsantomaggio do you want to test `pivotalrabbitmq/rabbitmq:stream-protocol-counters-otp-max` with https://github.com/rabbitmq/rabbitmq-stream-go-client/blob/main/examples/publishersError/publisherError.go ?

After that, if you want Gabriele, we can extend the new **RabbitMQ-Streams** dashboard with these new metrics 😉

**Comment by Gsantomaggio** [comment]
Sure !

**Comment by dcorbacho** [comment]
@gerhard Those test suites can't check for the new metrics, as they only show up when the stream protocol is used. There is a check for them on `rabbitmq_stream/test/rabbit_stream_SUITE`

**Comment by gerhard** [comment]
All looks good, starting QA in 3.9 LRE

**Comment by gerhard** [comment]
While this works, writing queries for these metrics is really difficult. This is what a single query looks like:

```
sum by(rabbitmq_cluster) (rabbitmq_global_stream_error_access_refused_total{protocol="stream"} * on(instance) group_left(rabbitmq_cluster, rabbitmq_node) rabbitmq_identity_info{rabbitmq_cluster="$rabbitmq_cluster", namespace="$namespace"}) +

sum by(rabbitmq_cluster) (rabbitmq_global_stream_error_authentication_failure_total{protocol="stream"} * on(instance) group_left(rabbitmq_cluster, rabbitmq_node) rabbitmq_identity_info{rabbitmq_cluster="$rabbitmq_cluster", namespace="$namespace"}) +

sum by(rabbitmq_cluster) (rabbitmq_global_stream_error_frame_too_large_total{protocol="stream"} * on(instance) group_left(rabbitmq_cluster, rabbitmq_node) rabbitmq_identity_info{rabbitmq_cluster="$rabbitmq_cluster", namespace="$namespace"}) +

sum by(rabbitmq_cluster) (rabbitmq_global_stream_error_internal_error_total{protocol="stream"} * on(instance) group_left(rabbitmq_cluster, rabbitmq_node) rabbitmq_identity_info{rabbitmq_cluster="$rabbitmq_cluster", namespace="$namespace"}) +

sum by(rabbitmq_cluster) (rabbitmq_global_stream_error_precondition_failed_total{protocol="stream"} * on(instance) group_left(rabbitmq_cluster, rabbitmq_node) rabbitmq_identity_info{rabbitmq_cluster="$rabbitmq_cluster", namespace="$namespace"}) +

sum by(rabbitmq_cluster) (rabbitmq_global_stream_error_publisher_does_not_exist_total{protocol="stream"} * on(instance) group_left(rabbitmq_cluster, rabbitmq_node) rabbitmq_identity_info{rabbitmq_cluster="$rabbitmq_cluster", namespace="$namespace"}) +

sum by(rabbitmq_cluster) (rabbitmq_global_stream_error_sasl_authentication_failure_loopback_total{protocol="stream"} * on(instance) group_left(rabbitmq_cluster, rabbitmq_node) rabbitmq_identity_info{rabbitmq_cluster="$rabbitmq_cluster", namespace="$namespace"}) +

sum by(rabbitmq_cluster) (rabbitmq_global_stream_error_sasl_challenge_total{protocol="stream"} * on(instance) group_left(rabbitmq_cluster, rabbitmq_node) rabbitmq_identity_info{rabbitmq_cluster="$rabbitmq_cluster", namespace="$namespace"}) +

sum by(rabbitmq_cluster) (rabbitmq_global_stream_error_sasl_error_total{protocol="stream"} * on(instance) group_left(rabbitmq_cluster, rabbitmq_node) rabbitmq_identity_info{rabbitmq_cluster="$rabbitmq_cluster", namespace="$namespace"}) +

sum by(rabbitmq_cluster) (rabbitmq_global_stream_error_sasl_mechanism_not_supported_total{protocol="stream"} * on(instance) group_left(rabbitmq_cluster, rabbitmq_node) rabbitmq_identity_info{rabbitmq_cluster="$rabbitmq_cluster", namespace="$namespace"}) +

sum by(rabbitmq_cluster) (rabbitmq_global_stream_error_stream_already_exists_total{protocol="stream"} * on(instance) group_left(rabbitmq_cluster, rabbitmq_node) rabbitmq_identity_info{rabbitmq_cluster="$rabbitmq_cluster", namespace="$namespace"}) +

sum by(rabbitmq_cluster) (rabbitmq_global_stream_error_stream_does_not_exist_total{protocol="stream"} * on(instance) group_left(rabbitmq_cluster, rabbitmq_node) rabbitmq_identity_info{rabbitmq_cluster="$rabbitmq_cluster", namespace="$namespace"}) +

sum by(rabbitmq_cluster) (rabbitmq_global_stream_error_stream_not_available_total{protocol="stream"} * on(instance) group_left(rabbitmq_cluster, rabbitmq_node) rabbitmq_identity_info{rabbitmq_cluster="$rabbitmq_cluster", namespace="$namespace"}) +

sum by(rabbitmq_cluster) (rabbitmq_global_stream_error_subscription_id_already_exists_total{protocol="stream"} * on(instance) group_left(rabbitmq_cluster, rabbitmq_node) rabbitmq_identity_info{rabbitmq_cluster="$rabbitmq_cluster", namespace="$namespace"}) +

sum by(rabbitmq_cluster) (rabbitmq_global_stream_error_subscription_id_does_not_exist_total{protocol="stream"} * on(instance) group_left(rabbitmq_cluster, rabbitmq_node) rabbitmq_identity_info{rabbitmq_cluster="$rabbitmq_cluster", namespace="$namespace"}) +

sum by(rabbitmq_cluster) (rabbitmq_global_stream_error_unknown_frame_total{protocol="stream"} * on(instance) group_left(rabbitmq_cluster, rabbitmq_node) rabbitmq_identity_info{rabbitmq_cluster="$rabbitmq_cluster", namespace="$namespace"}) +

sum by(rabbitmq_cluster) (rabbitmq_global_stream_error_vhost_access_failure_total{protocol="stream"} * on(instance) group_left(rabbitmq_cluster, rabbitmq_node) rabbitmq_identity_info{rabbitmq_cluster="$rabbitmq_cluster", namespace="$namespace"})
```

Could we capture the error type in a label, so that we would end up with metrics like these:

```
# TYPE rabbitmq_global_stream_error_total counter
# HELP rabbitmq_global_stream_error_total Total number of errors
rabbitmq_global_stream_error_total{protocol="stream",error_type="access_refused"} 0
rabbitmq_global_stream_error_total{protocol="stream",error_type="sasl_error"} 0
rabbitmq_global_stream_error_total{protocol="stream",error_type="authentication_failure"} 0
...

```

This would transform the above query into this:

```
sum by(rabbitmq_cluster, error_type) (rabbitmq_global_stream_error_total{protocol="stream"} * on(instance) group_left(rabbitmq_cluster, rabbitmq_node) rabbitmq_identity_info{rabbitmq_cluster="$rabbitmq_cluster", namespace="$namespace"})
```

As I mentioned above, while queries are difficult, this works and produces the following end-result:

image

**Comment by gerhard** [comment]
Having discussed this offline, it would be too difficult to capture the error types as labels. Leaving as is & merging. Thank you @dcorbacho!

**Comment by gerhard** [comment]
Backported to `v3.9.x`

**Pull request merged**